ProjFS.Mac/PrjFSKextTests/HandleVnodeOperationTests.mm

```diff
@@ -0,0 +1,514 @@
+#include "../PrjFSKext/PerformanceTracing.hpp"
+#include "../PrjFSKext/public/Message.h"
+#include "../PrjFSKext/ProviderMessaging.hpp"
+#include "../PrjFSKext/public/PrjFSXattrs.h"
+#import <XCTest/XCTest.h>
+#import <sys/stat.h>
+#include "KextMockUtilities.hpp"
+#include "MockVnodeAndMount.hpp"
+#include "MockProc.hpp"
+
+using std::shared_ptr;
+using std::vector;
+using KextMock::_;
+
+class PrjFSProviderUserClient
+{
+};
+
+bool ProviderMessaging_TrySendRequestAndWaitForResponse(
+    VirtualizationRootHandle root,
+    MessageType messageType,
+    const vnode_t vnode,
+    const FsidInode& vnodeFsidInode,
+    const char* vnodePath,
+    int pid,
+    const char* procname,
+    int* kauthResult,
+    int* kauthError)
+{
+    MockCalls::RecordFunctionCall(
+        ProviderMessaging_TrySendRequestAndWaitForResponse,
+        root,
+        messageType,
+        vnode,
+        vnodeFsidInode,
+        vnodePath,
+        pid,
+        procname,
+        kauthResult,
+        kauthError);
+    
+    return true;
+}
+
+static void SetFileXattrData(shared_ptr<vnode> vnode)
+{
+    PrjFSFileXAttrData rootXattr = {};
+    vector<uint8_t> rootXattrData(sizeof(rootXattr), 0x00);
+    memcpy(rootXattrData.data(), &rootXattr, rootXattrData.size());
+    vnode->xattrs.insert(make_pair(PrjFSFileXAttrName, rootXattrData));
+}
+
+@interface HandleVnodeOperationTests : XCTestCase
+@end
+
+@implementation HandleVnodeOperationTests
+{
+    vfs_context_t context;
+    const char* repoPath;
+    const char* filePath;
+    const char* dirPath;
+    PrjFSProviderUserClient dummyClient;
+    pid_t dummyClientPid;
+    shared_ptr<mount> testMount;
+    shared_ptr<vnode> repoRootVnode;
+    shared_ptr<vnode> testFileVnode;
+    shared_ptr<vnode> testDirVnode;
+}
+
+- (void) setUp
+{
+    kern_return_t initResult = VirtualizationRoots_Init();
+    XCTAssertEqual(initResult, KERN_SUCCESS);
+    context = vfs_context_create(NULL);
+    dummyClientPid = 100;
+
+    // Create Vnode Tree
+    repoPath = "/Users/test/code/Repo";
+    filePath = "/Users/test/code/Repo/file";
+    dirPath = "/Users/test/code/Repo/dir";
+    testMount = mount::Create();
+    repoRootVnode = testMount->CreateVnodeTree(repoPath, VDIR);
+    testFileVnode = testMount->CreateVnodeTree(filePath);
+    testDirVnode = testMount->CreateVnodeTree(dirPath, VDIR);
+
+    // Register provider for the repository path (Simulate a mount)
+    VirtualizationRootResult result = VirtualizationRoot_RegisterProviderForPath(&dummyClient, dummyClientPid, repoPath);
+    XCTAssertEqual(result.error, 0);
+    vnode_put(s_virtualizationRoots[result.root].rootVNode);
+}
+
+- (void) tearDown
+{
+    testMount.reset();
+    repoRootVnode.reset();
+    testFileVnode.reset();
+    testDirVnode.reset();
+    VirtualizationRoots_Cleanup();
+    vfs_context_rele(context);
+    MockVnodes_CheckAndClear();
+    MockCalls::Clear();
+}
+
+- (void) testEmptyFileHydrates {
```

Comment:
@wilbaker can you please do a quick scan of this file and let me know if you think we're missing any good cases.  I tried to test many relevant permutations.